Added message-delays option

Lemvs opened the pull request:
I decided to implement the option mentioned in this [Discussion](https://github.com/BG-Software-LLC/SuperiorSkyblock2/discussions/2730), but instead of creating another collection with delays only for `MISSION_NO_AUTO_REWARD` message, as is currently the case for `SCHEMATIC` and `PROTECTION` messages, I decided to allow players to create their own list of messages that should have delays.

By introducing this option, I can transfer the functionality of `protected-message-delay` option and add ability to set own delays for `SCHEMATIC`, `BUILD_OUTSIDE_ISLAND`, and `ISLAND_BEING_CALCULATED`. 

The only problem is that I can't use the methods I added in `PROTECTION` and `SCHEMATIC` messages, so maybe instead of creating such a custom message, I'll just send these messages normally? I'll simply move the `PROTECTION` functionality to `ProtectionHelper` and use the messages directly in `PortalsManagerServiceImpl` instead of through SCHEMATIC. And in the config, I would set by default the delays for these messages as they were before.

What do you think about this @OmerBenGera ?

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/core/messages/Message.java:
```diff
@@ -1079,6 +1085,27 @@ private void setMessage(Locale locale, IMessageComponent messageComponent) {
+        if (getDelay() > 0L) {
+            if (delays == null) {
+                delays = new HashMap<>();
+            }
```
delays should not be null. Also, delays should be ConcurrentHashMap

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/config/SettingsContainer.java:
```diff
@@ -562,6 +563,13 @@ public SettingsContainer(SuperiorSkyblockPlugin plugin, YamlConfiguration config
         tabCompleteHideVanished = config.getBoolean("tab-complete-hide-vanished", true);
         dropsUpgradePlayersMultiply = config.getBoolean("drops-upgrade-players-multiply", false);
         protectedMessageDelay = config.getLong("protected-message-delay", 60L);
```
No need for that anymore

Lemvs commented on src/main/java/com/bgsoftware/superiorskyblock/core/messages/Message.java:
```diff
@@ -1095,6 +1122,7 @@ public static void registerListeners(PluginEventsDispatcher dispatcher) {
 
     private static void onSettingsUpdate() {
         noInteractMessages = null;
+        delays = null;
```
Some time ago you did this for noInteractMessages, because after changing the delay in the config and reloading the plugin it wasn't updating, so I thought it was necessary to do the same here 🤔

Lemvs commented:
I also added a SCHEMATIC_NOT_ADDED message so that you can set a delay for it and translate it as well.
In the default config I entered all the messages that previously had delays.

But there is one problem, in SchematicManagerImpl in the convertData method I added converting the old protected-message-delay value, but it doesn't work - the values ​​are not changed, do you know why @OmerBenGera ?

I added it to IGNORED_SECTIONS, because without it, when loading the config, the deleted default messages from message-delays were coming back, I don't know if that has anything to do with it

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/core/messages/Message.java:
```diff
@@ -906,7 +849,7 @@ protected MessagesServiceImpl create() {
 
     @Nullable
-    private static Collection<UUID> noInteractMessages;
+    private static Map<Message, Collection<UUID>> delays;
```
You don't need this, you already have Message#delayedMessages (which can be renamed to just delays)

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/core/messages/Message.java:
```diff
@@ -1064,6 +1003,9 @@ public void send(CommandSender sender, Locale locale, Object... args) {
                 return;
+            if (addDelay(playerUUID)) {
+                return;
+            }
```
Redundant. You can use `delayedMessages`

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/core/messages/Message.java:
```diff
@@ -1079,6 +1021,28 @@ private void setMessage(Locale locale, IMessageComponent messageComponent) {
+
+        return false;
+    }
+
```
Can be removed probably

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/service/portals/PortalsManagerServiceImpl.java:
```diff
@@ -130,14 +129,14 @@ private EntityPortalResult handlePlayerPortalInternal(SuperiorPlayer superiorPla
 
-        if (portalResult == EntityPortalResult.DESTINATION_WORLD_DISABLED && !Message.WORLD_NOT_ENABLED.isEmpty(superiorPlayer.getUserLocale())) {
-            Message.SCHEMATICS.send(superiorPlayer, Message.WORLD_NOT_ENABLED.getMessage(superiorPlayer.getUserLocale()));
+        if (portalResult == EntityPortalResult.DESTINATION_WORLD_DISABLED) {
```
else if / switch

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/service/region/ProtectionHelper.java:
```diff
@@ -33,4 +46,26 @@ public static boolean shouldPreventInteraction(InteractionResult interactionResu
         throw new IllegalStateException("No handling for result " + interactionResult);
     }
 
+    public static void sendProtectionMessage(CommandSender sender) {
```
In most cases (if not all of them), Island was already looked-up by the caller.
Add a nullable island parameter to the method. Only if it's null, look for the island again.

OmerBenGera replied:
> But the shouldPreventInteraction() method is mainly used in ProtectionListener, where the island is not known yet, only in RegionManagerServiceImpl there are islands, so how to connect it? Should I "assign" the island to InteractionResult or how? 😕

In this case, pass null

OmerBenGera commented:
> I also added a SCHEMATIC_NOT_ADDED message so that you can set a delay for it and translate it as well. In the default config I entered all the messages that previously had delays.
> 
> But there is one problem, in SchematicManagerImpl in the convertData method I added converting the old protected-message-delay value, but it doesn't work - the values ​​are not changed, do you know why @OmerBenGera ?
> 
> I added it to IGNORED_SECTIONS, because without it, when loading the config, the deleted default messages from message-delays were coming back, I don't know if that has anything to do with it

Regarding convertData - I assume you meant SettingsManagerImpl. It should work, can't think about a reason why it wouldn't.
You need to add that section to the ignored sections, otherwise it will be synced with defaults. That's correct.

OmerBenGera commented on src/main/java/com/bgsoftware/superiorskyblock/core/messages/Message.java:
```diff
@@ -1094,7 +1029,19 @@ public static void registerListeners(PluginEventsDispatcher dispatcher) {
+    }
+
+    private long getDelay() {
+        Long delay = plugin.getSettings().getMessageDelays().get(this.name());
```
getOrDefault
